tephra-hbase-compat-0.94/src/main/java/co/cask/tephra/hbase94/TransactionAwareHTable.java

```diff
@@ -97,36 +91,72 @@ public TransactionAwareHTable(HTableInterface hTable, boolean allowNonTransactio
    * Create a transactional aware instance of the passed HTable, with the option
    * of allowing non-transactional operations.
    * @param hTable underlying HBase table to use
-   * @param conflictLevel level of conflict detection to perform
+   * @param conflictLevel level of conflict detection to perform (defaults to {@code COLUMN})
    * @param allowNonTransactional if true, additional operations (checkAndPut, increment, checkAndDelete)
    *                              will be available, though non-transactional
    */
   public TransactionAwareHTable(HTableInterface hTable, TxConstants.ConflictDetection conflictLevel,
                                 boolean allowNonTransactional) {
+    super(conflictLevel, allowNonTransactional);
     this.hTable = hTable;
-    this.changeSet = new ArrayList<ActionChange>();
-    this.txCodec = new TransactionCodec();
-    this.allowNonTransactional = allowNonTransactional;
-    this.conflictLevel = conflictLevel;
   }
 
-  /**
-   * True if the instance allows non-transaction operations.
-   * @return
-   */
-  public boolean getAllowNonTransactional() {
-    return this.allowNonTransactional;
+  /* AbstractTransactionAwareTable implementation */
+
+  @Override
+  protected byte[] getTableKey() {
+    return getTableName();
   }
 
-  /**
-   * Set whether the instance allows non-transactional operations.
-   * @param allowNonTransactional
-   */
-  public void setAllowNonTransactional(boolean allowNonTransactional) {
-    this.allowNonTransactional = allowNonTransactional;
+  @Override
+  protected boolean doCommit() throws IOException {
+    hTable.flushCommits();
+    return true;
   }
 
   @Override
+  protected boolean doRollback() throws Exception {
+    try {
+      List<Delete> rollbackDeletes = new ArrayList<Delete>(changeSet.size());
+      for (ActionChange change : changeSet) {
+        byte[] row = change.getRow();
+        byte[] family = change.getFamily();
+        byte[] qualifier = change.getQualifier();
+        long transactionTimestamp = tx.getWritePointer();
+        Delete rollbackDelete = new Delete(row, transactionTimestamp);
+        if (family != null && qualifier == null) {
```

Comment:
Yes, that is correct.  HBase 0.94 does not support a version-specific column family delete operation, which is what we use to rollback changes with row-level conflict detection.  So this retains the old behavior by keeping column level change sets for rollback, regardless of the conflict detection level used.

For HBase 0.96+, we only need to retain the row + family for each change, and use family deletes for rollback.